src/gt4py/next/backend.py

```diff
@@ -26,143 +27,138 @@
+        default_factory=past_to_itir.past_to_itir_factory
+    )
 
-DEFAULT_PROG_TRANSFORMS = ProgramTransformWorkflow()
+    def step_order(self, inp: INPUT_PAIR) -> list[str]:
+        steps: list[str] = []
+        if isinstance(inp.args, ARGS):
+            steps.append("aotify_args")
+        match inp.data:
+            case DSL_FOP():
+                steps.extend(
+                    [
+                        "func_to_foast",
+                        "field_view_op_to_prog",
+                        "past_lint",
+                        "field_view_prog_args_transform",
+                    ]
+                )
+            case FOP():
+                steps.extend(
+                    ["field_view_op_to_prog", "past_lint", "field_view_prog_args_transform"]
+                )
+            case DSL_PRG():
+                steps.extend(["func_to_past", "past_lint", "field_view_prog_args_transform"])
+            case PRG():
+                steps.extend(["past_lint", "field_view_prog_args_transform"])
+            case _:
+                pass
+        steps.append("past_to_itir")
+        return steps
+
+
+DEFAULT_TRANSFORMS: Transforms = Transforms()
 
 
 @dataclasses.dataclass(frozen=True)
 class Backend(Generic[core_defs.DeviceTypeT]):
-    executor: ppi.ProgramExecutor
+    executor: modular_executor.ModularExecutor
     allocator: next_allocators.FieldBufferAllocatorProtocol[core_defs.DeviceTypeT]
-    transforms_fop: FieldopTransformWorkflow = DEFAULT_FIELDOP_TRANSFORMS
-    transforms_prog: ProgramTransformWorkflow = DEFAULT_PROG_TRANSFORMS
+    transforms: workflow.Workflow[INPUT_PAIR, stages.AOTProgram]
 
     def __call__(
         self,
-        program: ffront_stages.ProgramDefinition | ffront_stages.FieldOperatorDefinition,
+        program: INPUT_DATA,
         *args: Any,
         **kwargs: Any,
     ) -> None:
-        if isinstance(
-            program, (ffront_stages.FieldOperatorDefinition, ffront_stages.FoastOperatorDefinition)
-        ):
-            offset_provider = kwargs.pop("offset_provider")
-            from_fieldop = kwargs.pop("from_fieldop")
-            program_call = self.transforms_fop(
-                workflow.InputWithArgs(program, args, kwargs | {"from_fieldop": from_fieldop})
-            )
-            program_call = dataclasses.replace(
-                program_call, kwargs=program_call.kwargs | {"offset_provider": offset_provider}
-            )
-        else:
-            program_call = self.transforms_prog(workflow.InputWithArgs(program, args, kwargs))
-        self.executor(program_call.program, *program_call.args, **program_call.kwargs)
+        if not isinstance(program, IT_PRG):
+            args, kwargs = signature.convert_to_positional(program, *args, **kwargs)
+        self.jit(program, *args, **kwargs)(*args, **kwargs)
+
+    def jit(self, program: INPUT_DATA, *args: Any, **kwargs: Any) -> stages.CompiledProgram:
```

Comment:
@samkellerhals Could you please review the new `.jit` and `.aot` methods whether they are useful enough for the `icon4py` usecase or if you see something missing (which we might implement now or in a follow-up PR).

Reply:
`.jit` is definitely useful in the context of the `CachedProgram` in icon4py. I am not sure what the intended use of `.aot` is as a standalone method, maybe you could help me understand. Also it would be useful if GT4Py had a class similar to the `CachedProgram` so that we can use that instead in icon4py. Another option could be to introduce a keyword argument in the `gtx.program` decorator e.g. `frozen=true`.

Reply:
`.aot` could be used, together with reading the argument types from the DSL type hints, to compile ahead-of-time. This will only become generally feasible if the offset provider is already given to the decorator (or, in future, the compile-time offset provider). Except in the cartesian case, where it could be done even now.

This would represent the absolute minimum of overhead, because no extra data structure is needed to trigger compilation on first call. Imagine in

```python
@gtx.program(precompile=True, offset_provider=...)
def foo(...) -> ...:
    ...
```

What you'd get stored as `foo` would already be the compiled program.

Reply:
Alternatively, if you have some "initialization step", that could call `backend.aot()` on all the programs.

Reply:
I added a `decorator.FrozenProgram` with an flag in `decorator.program`, as well as a `decorator.Program.freeze` method.